### DISCUSSION TITLE
Multiple async operations on the same session

### DISCUSSION THREAD

**Discussion by aventus13** [ravendb/ravendb#12320] [General]
I ran into a problem in my application. I have a fairly lightweight process where I execute two independent operations in parallel, like so:

```c#
var taskOne = DoSomethingAsync();
var taskTwo = DoSomethingElseAsync();

Task.WaitAll(taskOne, taskTwo);
```

Each task uses the same session to query data from the database. Please note that this is deliberate due to design of the application. However when running the code, it results in the following exception being thrown:

```
System.InvalidOperationException: Concurrent usage of async tasks in async session is forbidden, please make sure that async session doesn't execute async methods concurrently.
```

I would like to ask:

1. Why are concurrent async calls forbidden on the same session?
2. Is there a way to enable them? I can't find anything obvious on the session's conventions
3. If there is a way to enable them, what is the drawback of doing so, given that by default it is disabled?

**Comment by ayende** [comment]
A session isn't thread safe. If we allowed such concurrent operations, we would need to make all the session behavior thread safe.
That is hard and costly in terms of performance.

There is no way to enable this, for this scenario, use separate sessions. You can "import" an entity from another session using the `Store()` overload that accepts a change vector, etc, but change tracking will not work.

**Reply by mcrio** [reply]
> You can "import" an entity from another session using the Store() overload that accepts a change vector, etc, but change tracking will not work.

@aventus13 check if that requires your entities to be thread-safe

**Reply by aventus13** [reply]
Thanks for responses. In my case, my tasks don't do anything else than reading data independently- they also happen to read from the same collection, but it doesn't really matter. There is no transaction (saving of changes) involved at all, which is why I wanted to execute them in parallel in order to speed things up.

Given there is no way to enable this, I'm going to implement a work around- probably use await step by step. It's not ideal but as I said- the process is fairly lightweight so this is a slight inconvenience that I'm OK with.

When searching on Google Groups I came across a convention `AllowMultipleAsyncOperations`. @ayende is this no longer available in newer versions of Raven then?

**Reply by ayende** [reply]
That is no longer relevant, that is correct.
However, if they are independent operations, just use a separate session, and you'll have the same behavior.

**Reply by CyberCyclone** [reply]
@aventus13 I couldn't really tell what language you're using, but if it's NodeJS (others might be similar), you can use Promise.all() to run your queries run in parallel and then merge them once they're done.

**Reply by ayende** [reply]
@CyberCyclone From the `Task.WaitAll`, this is likely C#. And note that this isn't related to waiting on multiple promises. The issue is that RavenDB client directly blocks that scenario, because that would require us to assume that the session is used concurrently, which is a huge burden to take.

**Reply by loraderon** [reply]
I've encountered this issue as well. Are there any plans to address it in the future?

**Reply by ayende** [reply]
Please post a new issue, rather than opening a 4 years old thread
